example/nBody/src/nBody.cpp

```diff
@@ -0,0 +1,407 @@
+    {
+        if(!benchmarkMode)
+        {
+            std::cout << "================================" << std::endl;
+            std::cout << "Example n-Body simulation" << std::endl;
+            std::cout << "    Number of particles [#]: " << numParticles << std::endl;
+            std::cout << "    Time steps [#]: " << numTimeSteps << std::endl;
+            std::cout << "    dt: " << dt << std::endl;
+            if(writePngs)
+                std::cout << "    Writing pngs to disk" << std::endl;
+            else
+                std::cout << "    Not writing pngs to disk" << std::endl;
+            std::cout << "================================" << std::endl;
+            std::cout << std::endl;
+        }
+        else
+        {
+            // print csv header
+            std::cout << "# N-Body simulation, " << timeStepsBenchmark << " timesteps per run" << std::endl;
+            std::cout << "accelerator,numParticles,timePerStep(s),GFLOPs" << std::endl;
+        }
+    }
+
+    /** @brief Run an n-body simulation.
+     *
+     * Some number of particles are simulated in 3-dimensional space. Each particle gets a mass, position vector, and a
+     * velocity vector, initialized randomly (see common.hpp for constants to tweak the generation). The particles then
+     * interact gravitationally. The UpdateVelocitiesKernel updates each particle's velocity for one timestep and
+     * writes them back. The updated velocities only depend on the other particles' positions and masses, so there is
+     * no data race. Next, the UpdatePositionsKernel moves each particle independently according to its current
+     * velocity. This repeats for the desired number of time steps.
+     *
+     * @param deviceSpec The device specification to run on.
+     * @param computeExec The device to execute on.
+     * @param writePngs Whether to write pngs to disk or not.
+     * @param benchmarkMode Whether to run in benchmark mode. See the help docs below.
+     * @param numParticles The number of particles to simulate.
+     * @param numTimeSteps The number of time steps to run for.
+     * @param dt The delta t to use as time steps.
+     */
+    int example(
+        auto const deviceSpec,
+        auto const computeExec,
+        bool const writePngs,
+        bool const benchmarkMode,
+        IdxType const numParticles,
+        IdxType const numTimeSteps,
+        BaseType const dt)
+    {
+        using namespace alpaka;
+        using namespace alpaka::onHost;
+
+        using IdxTypeVec = Vec<IdxType, 1_idx>;
+
+        if(!benchmarkMode)
+            std::cout << "\nRunning accelerator: " << deviceSpec.getApi().getName() << std::endl;
+
+        auto devSelector = makeDeviceSelector(deviceSpec);
+        Device devAcc = devSelector.makeDevice(0);
+
+        // simulation defines
+        // {Y, X}
+        IdxTypeVec const extents = Vec{numParticles};
+
+        auto massesHost = onHost::allocHost<BaseType>(extents);
```

Comment:
should be `[[maybe_unused]] auto particleDataHost = ParticleData{` to avoid compielr warnings in case png is disabled